src/dmd/typesem.d

```diff
@@ -1299,7 +1299,8 @@ extern(C++) Type typeSemantic(Type type, const ref Loc loc, Scope* sc)
 
             // default arg must be an lvalue
             if (isRefOrOut && !isAuto &&
-                !(global.params.previewIn && (fparam.storageClass & STC.in_)))
+                !(global.params.previewIn && (fparam.storageClass & STC.in_)) &&
+                !(global.params.rvalueRefParam))
                 e = e.toLvalue(sc, e);
```

Comment:
Actually could you change this to
```d
if (isRefOrOut && !isAuto &&)
{
     if ((!(global.params.previewIn && (fparam.storageClass & STC.in_))) ||
                 !(global.params.rvalueRefParam))
    {
        e = e.toLvalue(sc, e);
        if (e.op == TOK.error)
          errorSupplimental(e.loc, "use `-preview=in` or `preview=rvaluerefparam`"
    }
}
```
note `||` not `&&`

Reply:
Are you sure about the `||` instead of `&&`?

Reply:
I've updated the condition again with the change to `||` along with and updated expected message for fail_compilation.